partner_type_base/models/res_partner.py

```diff
@@ -0,0 +1,46 @@
+# Copyright 2024 Christopher Rogos (<https://www.glueckkanja.com>)
+
+from odoo import api, fields, models
+
+
+class ResPartner(models.Model):
+    """Adds last name and first name; name becomes a stored function field."""
+
+    _inherit = "res.partner"
+
+    is_address_readonly = fields.Boolean(
+        compute="_compute_contact_type",
+        help="If true, the address fields are readonly.",
+    )
+    is_individual = fields.Boolean(
+        compute="_compute_contact_type", help="If true, the partner name is splitted."
+    )
+    can_be_parent = fields.Boolean(
+        compute="_compute_contact_type",
+        search="_search_can_be_parent",
+        help="If true, the partner is available as parent.",
+    )
+    can_be_child = fields.Boolean(
+        compute="_compute_contact_type",
+        help="If true, the partner_id field is available.",
+    )
+
+    @api.depends("is_company", "type", "parent_id")
+    def _compute_contact_type(self):
+        for partner in self:
+            partner.is_address_readonly = not (
+                partner.is_company
+                or not partner.parent_id
+                or partner.type not in ["contact"]
+            )
+            partner.is_individual = not partner.is_company and partner.type in [
+                "contact",
+                "other",
+            ]
+
+            partner.can_be_parent = partner.is_company
+            partner.can_be_child = not partner.is_company
+
+    @api.model
+    def _search_can_be_parent(self, operator, value):
+        return [("is_company", operator, value)]
```

Comment:
Is it real ? An individual can have a child partner as delivery address for instance.

Reply:
This is totally inconsistent in Odoo.

When you go to a contact, you can add a subcontact/address:
![image](https://github.com/user-attachments/assets/c6ed49b2-4839-40c0-8faa-deafc36071b5)

When you open this other address, the contact is set as parent: 
![image](https://github.com/user-attachments/assets/4401eb6e-089f-41cd-85d1-89d9bedee0c5)

But when you search for the contact to select it in this view, this is not possible:
![image](https://github.com/user-attachments/assets/5da67ce5-7782-41f6-8649-ce6f6efa6f2f)

In my Opinion a big weakness of Odoo, because it is very unclear to the user how to use it.
After setting a parent, you can also switch the contact to "company" and the parent_id is kept. 
Basically this could be good feature to address hierarchies, but I am not sure if all Odoo modules can work with this. 

What we are doing is, that we use a 3 level hirache Company -> "Other Address" -> Contact to bundle all people of a certain Office address. Because commerical_partner_id looks upwards until the first company, this should work. But I am not sure if it is a Odoo recommended structure. 
To have it more consistent, I added the can_be_parent attribute to make it more flexible and we overload the computation and add "Other Address" to the types which can_be_parent = True in another module.

Reply:
I had also opened an Odoo ticket on this topic 9 Months ago, and this was the answer:

> Thank you for your patience. I have been trying to find information on whether this behavior is supported or not, and in fact a lot of feedback has been created in the past suggesting we need this feature (three-level hierarchy with contacts). The reply the development team has given us in all of these cases is the behavior is indeed possible, but not entirely supported. Odoo does not support this behavior because it is a very problematic data structure, with a lot of hidden complexities, especially on the accounting aspect. This reply has been the same for feedbacks created 3 to 1 year ago, and the team does admit that if demand for this becomes very high it will be reconsidered.
> 
> Therefore, I have created a feedback ticket to shed light on the fact that you, like other customers, do need this hierarchy for your daily business flows. Depending on whether you just need to display that a contact belongs to another, the workaround proposed is to create a Studio field for it. If however this information will be used in other business flows such as accounting flows, custom code will be necessary to satisfy the need.